### PR TITLE
[FIL-405] + [FIL-404]

### PR DESCRIPTION
**Pull request by ortyomka**
Signed-off-by: ortyomka <iurin.art@gmail.com>

### CODE REVIEW

File: core/sector_storage/impl/scheduler_impl.cpp [hunk 1]
```diff
@@ -25,9 +47,79 @@ namespace fc::sector_storage {
...
+      const boost::optional<WorkId> &maybe_work_id) {
+    WorkerAction job = work;
+    ReturnCb callback = cb;
+    if (maybe_work_id.has_value()) {
+      auto &work_id{maybe_work_id.value()};
+
+      callback = [logger{logger_},
+                  old_cb = std::move(callback),
+                  kv{call_kv_},
+                  wid{maybe_work_id.value()}](
```
Comment (wer1st) [inline core/sector_storage/impl/scheduler_impl.cpp:60]:
```suggestion
                  wid{work_id}](
```

File: core/sector_storage/impl/scheduler_impl.cpp [hunk 2]
```diff
@@ -25,9 +47,79 @@ namespace fc::sector_storage {
...
       const ReturnCb &cb,
-      uint64_t priority) {
-    std::shared_ptr<TaskRequest> request = std::make_shared<TaskRequest>(
-        sector, task_type, priority, selector, prepare, work, cb);
+      uint64_t priority,
+      const boost::optional<WorkId> &maybe_work_id) {
+    WorkerAction job = work;
+    ReturnCb callback = cb;
+    if (maybe_work_id.has_value()) {
+      auto &work_id{maybe_work_id.value()};
```
Comment (wer1st) [inline core/sector_storage/impl/scheduler_impl.cpp:55]:
```suggestion
      const auto &work_id{maybe_work_id.value()};
```

File: core/sector_storage/impl/scheduler_impl.cpp [hunk 3]
```diff
@@ -25,9 +47,79 @@ namespace fc::sector_storage {
...
+      WorkState state;
+      state.id = work_id;
+      state.status = WorkStatus::kStart;
+      OUTCOME_TRY(state_raw, codec::cbor::encode(state));
+
+      OUTCOME_TRY(call_kv_->put(work_id, state_raw));
+
+      job = [old_job = std::move(job),
+             kv{call_kv_},
+             wid{maybe_work_id.value()}](const std::shared_ptr<Worker> &worker)
```
Comment (wer1st) [inline core/sector_storage/impl/scheduler_impl.cpp:102]:
```suggestion
             wid{work_id}](const std::shared_ptr<Worker> &worker)
```

File: core/sector_storage/scheduler.hpp [hunk 4]
```diff
@@ -15,6 +15,40 @@ namespace fc::sector_storage {
...
+      return result;
+    }
+  };
+  inline bool operator==(const WorkId &lhs, const WorkId &rhs) {
+    return lhs.task_type == rhs.task_type and lhs.param_hash == rhs.param_hash;
+  }
+
+  CBOR_TUPLE(WorkId, task_type, param_hash);
+
+  enum WorkStatus : uint64_t {
```
Comment (wer1st) [inline core/sector_storage/scheduler.hpp:37]: Use enum class

**Review by wer1st — APPROVED** [review]

**Review by turuslan — APPROVED** [review]

File: core/sector_storage/scheduler.hpp (hunk 4 above)
Comment (ortyomka) [reply core/sector_storage/scheduler.hpp]: Fixed

File: core/sector_storage/impl/scheduler_impl.cpp (hunk 3 above)
Comment (ortyomka) [reply core/sector_storage/impl/scheduler_impl.cpp]: Fixed

File: core/sector_storage/impl/scheduler_impl.cpp (hunk 2 above)
Comment (ortyomka) [reply core/sector_storage/impl/scheduler_impl.cpp]: Fixed

File: core/sector_storage/impl/scheduler_impl.cpp (hunk 1 above)
Comment (ortyomka) [reply core/sector_storage/impl/scheduler_impl.cpp]: Fixed

**Review by Alexey-N-Chernyshov — APPROVED** [review]

**Pull request merged**